### ISSUE TITLE
Page Should Contain took about 90 seconds

### ISSUE THREAD

**Issue by kanchi240** [robotframework/SeleniumLibrary#1156]
Hi, the keyword **Page Should Contain** took about 90 seconds in every execution, don't know why so slow, below is robot log screenshot:
![_20180716141354](https://user-images.githubusercontent.com/4908696/42747144-b14d5a26-890d-11e8-966d-50f7af7b0f85.png)

**Comment by aaltat** [comment]
Would you be able to run the test with `--loglevel trace` set from the command line and send the same screen capture (containing all the log messages) again? Also telling what browser, browser version, driver version (example chrome driver), selenium and what SeleniumLibrary versions you are using.

**Comment by kanchi240** [comment]
Hi,  aaltat, below is version info:
browser: chrome
browser version: 67.0.3396.99
chromedriver version: 2.40.565498
platform: Windows NT 10.0.17134 x86_64
selenium version: 3.12.0
SeleniumLibrary version: 3.1.1

and below is robot log with trace log level:
![_20180716161523](https://user-images.githubusercontent.com/4908696/42748659-89205c1e-8913-11e8-9868-53b58458090e.png)

**Comment by aaltat** [comment]
It looks like you have page that contains lots of frames. The `Page Should Contain` keyword looks also under frames (not all, but in some frames which is different story totally). But it looks like that each time SeleniumLibrary uses Selenium API to search an element, it just takes lot of time. We have test that tests the same thing, but our test page has [limited number of frames](https://github.com/robotframework/SeleniumLibrary/blob/master/atest/resources/html/frames/frameset.html) and elements and the search takes a fraction of a second.

I wonder have you encountered a bug in the Selenium (or Chrome driver implementation) and it performance is poor with pages uses lots of Frames? is you page available in public and would it possible to see it?

**Comment by kanchi240** [comment]
Yes, there are a lot of frames, and the text that need to be checked is in the last frame, and the text element's locator is different in different test case, so I use the **Page Should Contain** keyword to check the text:
![_20180716161523](https://user-images.githubusercontent.com/4908696/42751513-98a09424-891d-11e8-9583-a164357e2f25.png)

**Comment by pstefanowski** [comment]
I do not know if it will be helpful, but I always try to omit starting selector with //* because performance is bad, specially when element DOM is large. You need to write more specified xpath //frame[@id='fraInterface']//*[contains(.,'your_text')] or more accurate

**Comment by kanchi240** [comment]
@pstefanowski 
With SeleniumLibrary, which keyword should I use?

**Comment by kanchi240** [comment]
I try the **Current Frame Should Contain** keyword, but the element text is not in the page source(title attribute):
![1](https://user-images.githubusercontent.com/4908696/42754142-ef9ee084-8925-11e8-9cfa-5d49cb7e2210.png)
but in the chrome, the text is in the element(title attribute):
![2](https://user-images.githubusercontent.com/4908696/42754187-2773bfac-8926-11e8-9162-6a0d1034c19e.png)

**Comment by ombre42** [comment]
For each frame `Page Should Contain` searches in, it attempts to locate elements containing the search text. I suspect you have set the implicit timeout to 30 seconds. SeleniumLibrary searched and timed out in 3 frames before finding the element in the 4th frame. 3 x 30 seconds  = 90 seconds + a little overhead.

I see two solutions here:
* if the element (text) is not found in the root frame, set implicit wait to 0 before looking in any other frames
* set implicit wait to 0 before any element search

Pro to 1st solution: acknowledges that implicit wait is also generally used to help sync with the application under test so will be more likely to pass
Con to 1st solution: will wait the implicit wait as a minimum if the text is already present but in a sub-frame.

Pro to 2nd solution: speed
Con to 2nd solution: more likely to break existing tests

I am torn between these choices. Leaning towards the second until I think about what it is like to upgrade and find previously stable tests failing (possibly intermittently) as a result.

Either of these changes risk breaking tests. The recommendation for those who encounter this would be to switch to `Wait Until Page Contains`.

A more radical solution not mentioned above is getting the text of the body element of each frame (guaranteed to exist I believe). I presume there is a reason SL uses XPath and `contains` in lieu of this so only mentioning for completeness.

**Comment by kanchi240** [comment]
When I get the text element's `value` attribute(actually is `title` attribute)by specific locator, I could get the text value:
![1](https://user-images.githubusercontent.com/4908696/42792420-bf23877a-89a7-11e8-953b-37e16ccdd0d1.png)

but the text element in the page source has no `title` attribute or `value` attribute:
![2](https://user-images.githubusercontent.com/4908696/42792508-326413d0-89a8-11e8-93bf-462be1c73884.png)
![3](https://user-images.githubusercontent.com/4908696/42792514-37f1e570-89a8-11e8-937e-78851b688a99.png)
so I can't check by the text, because the text element locator is different in test cases that I can't check the text by specific locator.

**Comment by kanchi240** [comment]
Yes, I set implicit wait to 30 seconds when creating the driver.
I set implicit wait to 0 before `Page Should Contain`, then the test passed very quickly:
![1](https://user-images.githubusercontent.com/4908696/42798883-56d35a9e-89c8-11e8-8984-5317a26bcd65.png)

In the test steps, I get the text element `value` attribute value and **check the value is not empty firstly**, then check the text content with:

- `Wait Until Page Contains`
- `Wait Until Page Contains Element`
- `Page Should Contain Element`

all these 3 keywords have the similar failed message as below:
![2](https://user-images.githubusercontent.com/4908696/42800728-91675910-89cf-11e8-85bc-9ee58c2b4ada.png)



and the text element `value` attribute is not present in the page source:
![3](https://user-images.githubusercontent.com/4908696/42801200-5bc2c8ba-89d1-11e8-8109-9e3c5eb94edc.png)

problem:

- Why `Page Should Contain` success but other 3 keywords above failed(no matter whether setting implicit wait to 0 or not)? see update below.
- Get the element attribute success firstly, but why attribute is not present in the page source when calling `Wait Until Page Contains Element` or `Page Should Contain Element`?


update:
The search text part in the code of `Wait Until Page Contains` and `Page Should Contain` is different:

- `Wait Until Page Contains`: just search the text in the current frame
- `Page Should Contain`: lookup every frame
